File: AlerteTCL/Services/WikimediaService.swift
import Foundation

/// Recherche des photos de véhicules sur Wikimedia Commons (CC-BY-SA).
/// API publique, sans clé. Politique d'usage : https://www.mediawiki.org/wiki/API:Etiquette
enum WikimediaService {
    private static let baseURL = "https://commons.wikimedia.org/w/api.php"
    private static var cache: [String: [URL]] = [:]

    /// Retourne les URLs de thumbnails (400 px) pour le modèle donné, triées par index.
    static func fetchPhotos(for model: String) async -> [URL] {
        if let cached = cache[model] { return cached }

        let query = "\(model) TCL"
        guard let encoded = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
              let url = URL(string: "\(baseURL)?action=query&generator=search&gsrnamespace=6&gsrsearch=\(encoded)&gsrlimit=8&prop=imageinfo&iiprop=url&iiurlwidth=400&format=json") else {
            return []
        }

        var request = URLRequest(url: url, timeoutInterval: 8)
        // Wikimedia API policy : User-Agent obligatoire avec contact
        request.setValue("AlerteTCL/1.0 (iOS; contact@alertetcl.fr)", forHTTPHeaderField: "User-Agent")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else { return [] }
            let result = try JSONDecoder().decode(WikimediaResponse.self, from: data)
            guard let pages = result.query?.pages else {
                cache[model] = []
                return []
            }
            let urls = pages.values
                .sorted { ($0.index ?? Int.max) < ($1.index ?? Int.max) }
                .compactMap { $0.imageinfo?.first?.thumburl }
                .compactMap { URL(string: $0) }
            cache[model] = urls
            return urls
        } catch {
            return []
        }
    }

    // MARK: - DTOs

    private struct WikimediaResponse: Decodable {
        let query: WikimediaQuery?
    }

    private struct WikimediaQuery: Decodable {
        let pages: [String: WikimediaPage]
    }

    private struct WikimediaPage: Decodable {
        let index: Int?
        let imageinfo: [WikimediaImageInfo]?
    }

    private struct WikimediaImageInfo: Decodable {
        let thumburl: String?
    }
}

File: shared/src/commonMain/kotlin/com/alertetcl/shared/services/WikimediaService.kt
package com.alertetcl.shared.services

import com.alertetcl.shared.network.HttpClientProvider
import io.ktor.client.call.body
import io.ktor.client.plugins.timeout
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.http.HttpStatusCode
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable

/**
 * Recherche des photos de véhicules sur Wikimedia Commons (CC-BY-SA).
 * API publique, sans clé. Politique d'usage : https://www.mediawiki.org/wiki/API:Etiquette
 */
class WikimediaService {
    private val client = HttpClientProvider.client
    private val baseURL = "https://commons.wikimedia.org/w/api.php"

    private val mutex = Mutex()
    private val cache = mutableMapOf<String, List<String>>()

    /** Retourne les URLs de thumbnails (400 px) pour le modèle donné, triées par index. */
    suspend fun fetchPhotoUrls(model: String): List<String> {
        mutex.withLock { cache[model] }?.let { return it }
        val query = "$model TCL"
        return try {
            val resp = client.get(baseURL) {
                url {
                    parameters.append("action", "query")
                    parameters.append("generator", "search")
                    parameters.append("gsrnamespace", "6")
                    parameters.append("gsrsearch", query)
                    parameters.append("gsrlimit", "8")
                    parameters.append("prop", "imageinfo")
                    parameters.append("iiprop", "url")
                    parameters.append("iiurlwidth", "400")
                    parameters.append("format", "json")
                }
                header("User-Agent", "AlerteTCL/1.0 (Android; contact@alertetcl.fr)")
                timeout { requestTimeoutMillis = 8_000L }
            }
            if (resp.status != HttpStatusCode.OK) return emptyList()
            val result = resp.body<WikimediaResponse>()
            val urls = result.query?.pages?.values
                ?.sortedBy { it.index ?: Int.MAX_VALUE }
                ?.mapNotNull { it.imageinfo?.firstOrNull()?.thumburl }
                ?: emptyList()
            mutex.withLock { cache[model] = urls }
            urls
        } catch (_: Throwable) {
            emptyList()
        }
    }

    // MARK: - DTOs

    @Serializable
    private data class WikimediaResponse(val query: WikimediaQuery? = null)

    @Serializable
    private data class WikimediaQuery(val pages: Map<String, WikimediaPage>? = null)

    @Serializable
    private data class WikimediaPage(
        val index: Int? = null,
        val imageinfo: List<WikimediaImageInfo>? = null
    )

    @Serializable
    private data class WikimediaImageInfo(val thumburl: String? = null)

    companion object { val shared = WikimediaService() }
}
